### DISCUSSION TITLE
Easy way to include OrderNumber in PDF-Export Summary (like "projectName (orderNumber)")

### DISCUSSION THREAD

**Discussion by skipperTux** [kimai/kimai#2188] [Questions & Answers] [answered]
Hi. I would like to include the Project OrderNumber in the PDF-Export. With the help of the [Export Documentation](https://www.kimai.org/documentation/export.html#adding-export-templates) it was easy to add a new template and check if changing the template works 👍. I also found the correct line to add the orderNumber, below line 168 after {{ summary.project }}. However when I added these lines of code (with regards to some similar code in the invoice template) it did not work out.
```
                  {% if summary.project.orderNumber is not empty %}
                    ({{ summary.project.orderNumber }})
                  {% endif %}
```
Looking into [Export/Base/PDFRenderer.php](https://github.com/kevinpapst/kimai2/blob/33f76f426c9e48a39089775afc4459206b2ce7f0/src/Export/Base/PDFRenderer.php) and the [calculateSummary](https://github.com/kevinpapst/kimai2/blob/7a3388646b46fe9758b905f390e8fbe67287863a/src/Export/Base/RendererTrait.php) function I learned that the Summary Class does not have the orderNumber property.
What is the best way to proceed? Add a new PDFRenderer.php based on a new RendererTrait.php in scr/Export/Base? The [Adding export renderer](https://www.kimai.org/documentation/export.html#adding-export-renderer) Documentation is not going into detail, and I only have limited php knowledge and no Symfony Framework knowledge. Maybe I am missing an easier way.
Thanks in advance.

**Comment by kevinpapst** [comment, marked as answer]
Good research! But there is no need to patch the PHP, you should be able to do everything in twig.
You can iterate above the entries with a twig foreach and fill an array with projectId = order number.
That can be used to include the order number in the summary.

Something like that in the beginning in the template:
```
    {% set projectToOrderNumber = {} %}
    {% for entry in entries %}
        {% set projectToOrderNumber = projectToOrderNumber |merge({('#' ~ entry.project.id): entry.project.orderNumber}) %}
    {% endfor %}
```
and then in the summary:
```
    {% for id, summary in summaries %}
        {# ... #}
        {{ projectToOrderNumber['#' ~ id] }}
        {# ... #}
    {% endfor %}
```
There is something weird going on with Twig and integer indices in arrays, so the workaround with the prepended `#`


Edit: untested code, only to give you an idea!

**Reply by skipperTux** [reply]
Thanks for pointing me in the right direction. For reference, my changes for a working example.

In the [calculateSummary](https://github.com/kevinpapst/kimai2/blob/7a3388646b46fe9758b905f390e8fbe67287863a/src/Export/Base/RendererTrait.php) function the id is defined as
```
$id = $customerId . '_' . $projectId;
```
so some splitting is needed to get the Project-Id. Second, only add Project-Ids that have an OrderNumber to avoid empty parentheses.

Put filling the array with key-value-pairs (Project-Id : Project-OrderNumber) somewhere in the beginning of the template. (Line 135, where all vars are initialized in &lt;tbody&gt; is a good place.)
```
        <tbody>
        {# ... #}
        {% set projectOrderNumbers = {} %}
        {% for entry in entries %}
            {% set projectId = ('#' ~ entry.project.id) %}
            {% if (projectId not in projectOrderNumbers|keys) and (entry.project.orderNumber is not empty) %}
                {% set projectOrderNumbers = projectOrderNumbers|merge({ (projectId) : entry.project.orderNumber }) %}
            {% endif %}
        {% endfor %}
```
Put the OrderNumber snippet into the Project-summary cell (new line 174).
```
                <td>
                    {{ summary.project }}
                    {% set projectId = ('#' ~ (id|split('_', 2))[1] ) %}
                    {% if projectId in projectOrderNumbers|keys %}
                        ({{ projectOrderNumbers[projectId] }})
                    {% endif %}
                </td>
```
